src/data/air_composition.F90

```diff
@@ -510,26 +515,71 @@ end subroutine air_composition_init
 
-   subroutine air_composition_update(mmr, ncol, to_moist_factor)
+   subroutine dry_air_composition_update(mmr, ncol, to_dry_factor)
 
-      real(kind_phys),           intent(in) :: mmr(:,:,:) ! constituents array
+      !(mmr = dry mixing ratio, if not, use to_dry_factor to convert!)
+      real(kind_phys),           intent(in) :: mmr(:,:,:) ! mixing ratios for species dependent dry air
       integer,                   intent(in) :: ncol       ! number of columns
-      real(kind_phys), optional, intent(in) :: to_moist_factor(:,:)
+      real(kind_phys), optional, intent(in) :: to_dry_factor(:,:)
 
       call get_R_dry(mmr(:ncol, :, :), thermodynamic_active_species_idx, &
-           rairv(:ncol, :), fact=to_moist_factor)
+           rairv(:ncol, :), fact=to_dry_factor)
       call get_cp_dry(mmr(:ncol,:,:), thermodynamic_active_species_idx, &
-           cpairv(:ncol,:), fact=to_moist_factor)
+           cpairv(:ncol,:), fact=to_dry_factor)
       call get_mbarv(mmr(:ncol,:,:), thermodynamic_active_species_idx, &
-           mbarv(:ncol,:), fact=to_moist_factor)
+           mbarv(:ncol,:), fact=to_dry_factor)
 
       cappav(:ncol,:) = rairv(:ncol,:) / cpairv(:ncol,:)
 
-   end subroutine air_composition_update
+   end subroutine dry_air_composition_update
+
+   !===========================================================================
+   !---------------------------------------------------------------------------
+   ! water_composition_update: Update generalized cp or cv depending on dycore
+   !---------------------------------------------------------------------------
+   !===========================================================================
+   subroutine water_composition_update(mmr, ncol, energy_formula, to_dry_factor)
+      use cam_thermo_formula, only: ENERGY_FORMULA_DYCORE_FV, ENERGY_FORMULA_DYCORE_SE, ENERGY_FORMULA_DYCORE_MPAS
+
+      real(kind_phys),           intent(in) :: mmr(:,:,:)     ! constituents array
+      integer,                   intent(in) :: ncol           ! number of columns
+      integer,                   intent(in) :: energy_formula ! energy formula for dynamical core
+      real(kind_phys), optional, intent(in) :: to_dry_factor(:,:)
+
+      character(len=*), parameter :: subname = 'water_composition_update'
+
+      ! update enthalpy or internal energy scaling factor for energy consistency with CAM physics
+      ! cp_or_cv_dycore is now a registry variable (physics_types) in CAM-SIMA instead of in-module
+
+      if (energy_formula == ENERGY_FORMULA_DYCORE_FV) then
+         ! FV: moist pressure vertical coordinate does not need update.
+      else if (energy_formula == ENERGY_FORMULA_DYCORE_SE) then
+         ! SE
+
+         ! **TEMP** TODO hplin 9/17/24:
+         ! for compatibility with CAM-SIMA that allocates thermodynamic_active_species_idx(0:num_advected)
+         ! (whereas CAM only allocates 1-indexed) I subset it here. This should be verified during code
+         ! review.
```

Comment:
Thanks! I'm not sure what the policy on 'todo' comments are and if I can reference a issue number, but I modified it so it's short while still noting the purpose of the `(1:)` subsetting. Or, I'm happy to quote the code in #334 as a reminder. Please let me know what you prefer!

```
! Note: species index subset to 1: because SIMA currently uses index 0. See #334.
```